google/data_source_google_service_account_key.go

```diff
@@ -67,3 +70,30 @@ func dataSourceGoogleServiceAccountKeyRead(d *schema.ResourceData, meta interfac
 
 	return nil
 }
+
+func getDataSourceServiceAccountKeyName(d *schema.ResourceData) (string, error) {
+	r := regexp.MustCompile("projects/(.+)/serviceAccounts/(.+)/keys/(.+)")
+
+	keyName := d.Get("name").(string)
+	keyFromSAId := d.Get("service_account_id").(string)
+
+	// Neither name nor service_account_id specified
+	if keyName == "" && keyFromSAId == "" {
+		return "", fmt.Errorf("please use name to specify service account key being added as this data source")
+	}
+	// Both name and service_account_id specified
+	if keyName != "" && keyFromSAId != "" {
+		return "", fmt.Errorf("please do not use both name and deprecated service_account_id fields")
+	}
+
+	fullKeyName := keyName
+	// Key name specified as incorrectly named, deprecated service account ID field
+	if keyName == "" {
+		fullKeyName = keyFromSAId
+	}
+
+	if r.MatchString(fullKeyName) {
```

Comment:
Oh, what about adding a ValidateFunc to "name" so these are caught plan-time? That would let us just return the name a bit early from this if it's set.

Reply:
I'm going add ValidateFunc to the schema but also leave this here since interpolated values don't trigger ValidateFunc.